FWCore/Framework/src/Worker.cc

```diff
@@ -267,63 +268,65 @@ namespace edm {
       tbb::task_arena edArena(tbb::task_arena::attach{});
-      tbb::task::spawn(
-          *make_functor_task(tbb::task::allocate_root(),
-                             [this, task = edm::WaitingTaskHolder(iTask), iTrans, &iImpl, iToken, edArena]() mutable {
-                               esTaskArena().execute([this, iTrans, &iImpl, iToken, task = std::move(task), edArena]() {
-                                 auto waitTask = edm::make_empty_waiting_task();
-                                 auto const& recs = esRecordsToGetFrom(iTrans);
-                                 auto const& items = esItemsToGetFrom(iTrans);
-                                 waitTask->set_ref_count(2);
-                                 for (size_t i = 0; i != items.size(); ++i) {
-                                   if (recs[i] != ESRecordIndex{}) {
-                                     auto rec = iImpl.findImpl(recs[i]);
-                                     if (rec) {
-                                       rec->prefetchAsync(waitTask.get(), items[i], &iImpl, iToken);
-                                     }
-                                   }
-                                 }
-                                 waitTask->decrement_ref_count();
-                                 waitTask->wait_for_all();
-
-                                 auto exPtr = waitTask->exceptionPtr();
-                                 tbb::task_arena(edArena).execute([task, exPtr]() {
-                                   auto t = task;
-                                   if (exPtr) {
-                                     t.doneWaiting(*exPtr);
-                                   } else {
-                                     t.doneWaiting(std::exception_ptr{});
-                                   }
-                                 });
-                               });
-                             }));
+      tbb::task::spawn(*make_functor_task(
+          tbb::task::allocate_root(), [this, task = std::move(iTask), iTrans, &iImpl, iToken, edArena]() mutable {
+            esTaskArena().execute([this, iTrans, &iImpl, iToken, task = std::move(task), edArena]() {
+              auto waitTask = edm::make_empty_waiting_task();
+              auto const& recs = esRecordsToGetFrom(iTrans);
+              auto const& items = esItemsToGetFrom(iTrans);
+              waitTask->set_ref_count(2);
+              {
+                //Need hWaitTask to go out of scope before calling wait_for_all
+                WaitingTaskHolder hWaitTask(waitTask.get());
+                for (size_t i = 0; i != items.size(); ++i) {
+                  if (recs[i] != ESRecordIndex{}) {
+                    auto rec = iImpl.findImpl(recs[i]);
+                    if (rec) {
+                      rec->prefetchAsync(hWaitTask, items[i], &iImpl, iToken);
+                    }
+                  }
+                }
+              }
+              waitTask->decrement_ref_count();
```

Comment:
How useful is the
```cpp
waitTask->set_ref_count(2);
{ ... }
waitTask->decrement_ref_count();
waitTask->wait_for_all();
```
compared to
```cpp
waitTask->set_ref_count(1);
{ ... }
waitTask->wait_for_all();
```
when `WaitingTaskHolder` constructor increases the reference count of the task?

Reply:
That was a left over from the conversion and isn't needed anymore. On the other hand, it doesn't really hurt either and will be completely replaced when we do the conversion to `tbb::task_group`.